os/kernel/sched/sched_migrate_tasks.c

```diff
@@ -0,0 +1,152 @@
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <tinyara/config.h>
+
+#ifdef CONFIG_SCHED_MIGRATE
+
+#include <stdbool.h>
+#include <errno.h>
+#include <debug.h>
+
+#include <tinyara/arch.h>
+#include <tinyara/spinlock.h>
+#include <tinyara/sched.h>
+
+#include "sched/sched.h"
+#include "clock/clock.h"
+
+
+/****************************************************************************
+ * Name: sched_migrate_tasks
+ *
+ * Description:
+ *   Migrates all tasks from a specified offline CPU to other available
+ *   online CPUs.
+ *
+ *   IMPORTANT: This function must be called from a running CPU (e.g., CPU0)
+ *   The caller must ensure preemption is locked or this is called
+ *   in a context where it cannot be preempted.
+ *
+ * Inputs:
+ *   offline_cpu - The index of the CPU that needs to be shut down.
+ *
+ * Return Value:
+ *   OK (0) on success, or a negated errno value on failure.
+ *
+ ****************************************************************************/
+
+int sched_migrate_tasks(int offline_cpu)
+{
+	FAR struct tcb_s *tcb;
+	FAR struct tcb_s *next_tcb;
+	FAR dq_queue_t *task_list;
+	irqstate_t flags;
+	int ret = OK;
+
+	/* Basic sanity checks */
+	if (offline_cpu < 0 || offline_cpu >= CONFIG_SMP_NCPUS) {
+		lldbg("ERROR: Invalid offline_cpu: %d\n", offline_cpu);
+		return -EINVAL;
+	}
+
+	if (offline_cpu == this_cpu()) {
+		lldbg("ERROR: Cannot migrate tasks from the current running CPU: %d\n", offline_cpu);
+		return -EINVAL;
+	}
+
+	/* Lock the scheduler to prevent preemption during this operation.
+	 * This ensures that no other tasks can be scheduled while we're
+	 * migrating tasks from the offline CPU.
+	 */
+	sched_lock();
+
+	/* Disable interrupts to ensure atomicity throughout this critical operation. */
+	flags = enter_critical_section();
+
+	/* Update the active CPUs mask to indicate that the offline CPU is no longer active */
+	CPU_CLR(offline_cpu, &g_active_cpus_mask);
+
+	DEBUGVERIFY(up_cpu_pause(offline_cpu));
+
+	/* Get the list of tasks assigned to the offline CPU.
+	 * We need to lock the task list to prevent concurrent access, though
+	 * since offline_cpu is halted, this might be redundant but is safe.
+	 */
+
+	task_list = (FAR dq_queue_t *)&g_assignedtasks[offline_cpu];
+
+	/* Iterate through each task in the offline CPU's assigned list.
+	 * Since there's no _foreach_safe macro, we must manually get the next
+	 * pointer before removing the current TCB.
+	 */
+
+	tcb = (FAR struct tcb_s *)task_list->head;
+	while (tcb != NULL) {
+		next_tcb = (FAR struct tcb_s *)tcb->flink;	/* Get next TCB before removing current one */
```

Comment:
```suggestion
	tcb = (FAR struct tcb_s *)task_list->tail;
	while (tcb != NULL) {
		priv_tcb = (FAR struct tcb_s *)tcb->blink;	/* Get next TCB before removing current one */
```

If you do not want to repeatedly change the head task (running task),
Let's change the order of progress of this to the opposite.

not head task removing or adding is much simple